clarity/src/vm/analysis/type_checker/tests/contracts.rs

```diff
@@ -630,3 +664,1864 @@ fn test_expects() {
+
+        let trait_with_compatible_trait_arg = "(define-trait trait-1 (
+            (echo (uint) (response uint uint))
+        ))
+        (define-trait trait-2 (
+            (echo (uint) (response uint uint))
+        ))
+        (define-trait trait-a (
+            (echo (<trait-1>) (response uint uint))
+        ))
+        (define-trait trait-b (
+            (echo (<trait-2>) (response uint uint))
+        ))
+        (define-public (wrapped-echo (contract <trait-a>))
+            (internal-echo contract))
+        (define-public (internal-echo (contract <trait-b>) (callee <trait-2>))
+            (contract-call? contract echo callee))";
+
+        let trait_with_bad_trait_arg = "(define-trait trait-1 (
+            (echo (uint) (response uint uint))
+        ))
+        (define-trait trait-2 (
+            (echo (uint) (response int uint))
+        ))
+        (define-trait trait-a (
+            (echo (<trait-1>) (response uint uint))
+        ))
+        (define-trait trait-b (
+            (echo (<trait-2>) (response uint uint))
+        ))
+        (define-public (wrapped-echo (contract <trait-a>))
+            (internal-echo contract))
+        (define-public (internal-echo (contract <trait-b>) (callee <trait-2>))
+            (contract-call? contract echo callee))";
+
+        let trait_with_superset_trait_arg = "(define-trait trait-1 (
+            (echo (uint) (response uint uint))
+        ))
+        (define-trait trait-2 (
+            (echo (uint) (response uint uint))
+            (foo (uint) (response uint uint))
+        ))
+        (define-trait trait-a (
+            (echo (<trait-1>) (response uint uint))
+        ))
+        (define-trait trait-b (
+            (echo (<trait-2>) (response uint uint))
+        ))
+        (define-public (wrapped-echo (contract <trait-a>))
+            (internal-echo contract))
+        (define-public (internal-echo (contract <trait-b>) (callee <trait-2>))
+            (contract-call? contract echo callee))";
+
+        let trait_with_subset_trait_arg = "(define-trait trait-1 (
+            (echo (uint) (response uint uint))
+        ))
+        (define-trait trait-2 (
+            (echo (uint) (response uint uint))
+            (foo (uint) (response uint uint))
+        ))
+        (define-trait trait-a (
+            (echo (<trait-1>) (response uint uint))
+        ))
+        (define-trait trait-b (
+            (echo (<trait-2>) (response uint uint))
+        ))
+        (define-public (wrapped-echo (contract <trait-b>))
+            (internal-echo contract))
+        (define-public (internal-echo (contract <trait-a>) (callee <trait-1>))
+            (contract-call? contract echo callee))";
+
+        let trait_with_duplicate_method = "(define-trait double-method (
+            (foo (uint) (response uint uint))
+            (foo (bool) (response bool bool))
+          ))";
+
+        let trait_to_subtrait_and_back = "(define-trait trait-1 (
```

Comment:
Wait, does this work?  Should this work?

Reply:
No, this does not and should not work.